src/http-api/resources/files.js

```diff
@@ -0,0 +1,82 @@
+'use strict'
+
+const bs58 = require('bs58')
+const ndjson = require('ndjson')
+const Readable = require('stream').Readable
+const multipart = require('ipfs-multipart')
+const debug = require('debug')
+const log = debug('http-api:files')
+log.error = debug('http-api:files:error')
+
+exports = module.exports
+
+exports.add = {
+  handler: (request, reply) => {
+    if (!request.payload) {
+      return reply('Array, Buffer, or String is required').code(400).takeover()
+    }
+
+    const parser = multipart.reqParser(request.payload)
+
+    var file = false
+    var written = 0
+
+    var serialize = ndjson.serialize()
+    // hapi doesn't permit object streams: http://hapijs.com/api#replyerr-result
+    serialize._readableState.objectMode = false
+
+    var fileAdder = request.server.app.ipfs.files.add()
+
+    fileAdder.on('data', (file) => {
+      serialize.write({
+        Name: file.path,
+        Hash: bs58.encode(file.multihash).toString()
+      })
+      written++
+    })
+
+    fileAdder.on('end', () => {
+      if (written === 0 && file) {
```

Comment:
if the ```file``` flag is true, but the ```written``` var hasn't been incremented, then a 500 server error is returned because the core failed even though the request had valid data parsed.

If at the end of parsing the request, if ```file``` is still false then the request was bad, so ```400 Bad Request``` is returned